dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+  }
+  TensorShape(std::array<int64_t, dim_> &&s) : shape(std::move(s)) {
+    std::cout << "std::array<int64_t, dim_> &&s" << std::endl;
+  }
+
+  // We just want zero-initialization
+  TensorShape() : shape{} {}
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {
+    static_assert(sizeof...(Ts) == dim_ - 1, "Number of shapes passed must match dim_");
+    std::cout << "TensorShape(int64_t i0, Ts... s)" << std::endl;
+    // TODO static_assert for Ts == int64_t
+  }
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::array<int64_t, dim_> shape;
+  constexpr int size() const { return dim_; }
+};
+
+// template<int other_dim>
+// TensorShape<DynamicTensorShape>::operator TensorShape<other_dim>() const {
+//   std::cout << "operator TensorShape<other_dim>()" << std::endl;
+//   assert(size() == other_dim);
+//   TensorShape<other_dim> shape;
+//   for (int i = 0; i < other_dim; i++) {
+//     shape[i] = (*this)[i];
+//   }
+//   return shape;
+// }
+
+template <int other_dim>
+TensorShape<other_dim> TensorShape<DynamicTensorShape>::to_static() const {
+  std::cout << "operator TensorShape<other_dim>()" << std::endl;
+  assert(size() == other_dim);
+  TensorShape<other_dim> shape;
+  for (int i = 0; i < other_dim; i++) {
+    shape[i] = (*this)[i];
+  }
+  return shape;
+}
+
+// TODO should we be able to only output a tensor of sample_dim dim for all cases of TensorListShape
+template <int sample_dim_>
+struct TensorListShape {
+  TensorListShape(std::vector<TensorShape<sample_dim_>> sample_shapes)
+      : shapes(std::accumulate(sample_shapes.begin(), sample_shapes.end(), std::vector<int64_t>(),
+                               [](std::vector<int64_t> current, const TensorShape<sample_dim_> &next) {
+                                 return std::move(current).insert(current.end(), next.shape.begin(),
+                                                                  next.shape.end());
+                               })) {}
+
+  template <int tensor_dim = sample_dim_>
+  typename std::enable_if<tensor_dim == sample_dim_ || tensor_dim == DynamicTensorShape,
+                          TensorShape<tensor_dim>>::type
+  tensor_shape(int64_t sample) const {
+    TensorShape<tensor_dim> out;
+    if (tensor_dim == DynamicTensorShape) {
+      out.shape.resize(sample_dim());
+    }
+    int64_t base = sample_dim() * sample;
+    for (int i = 0; i < sample_dim(); i++) {
+      out[i] = shapes[base + i];
+    }
+    return out;
+  }
+
+  span<int64_t, sample_dim_> tensor_shape_span(int64_t sample) const {
+    return {&shapes[sample * sample_dim()], sample_dim()};
```

Comment:
The second argument is not necessary - it'll only need to get checked, if provided.

Reply:
Done.